### PR TITLE
Database Writers & Configuration/Factory Refactor

### PR DESCRIPTION
**Pull request by ryanhassing**
Database objects capable of writing one row, or multiple rows, to a database file.

### CODE REVIEW

File: src/chariot/api/network/NetworkManagerAPI.py [hunk 1]
```diff
@@ -1,22 +1,47 @@
 from typing import List
 from flask import Flask, request, app, jsonify, request
+<<<<<<< HEAD
```
Comment (enioluwa23) [inline src/chariot/api/network/NetworkManagerAPI.py]: Can you fix all the merge conflicts in this file?

File: src/chariot/database/Database.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import abc
+from typing import List
+
+
+class Database(metaclass=abc.ABCMeta):
```
Comment (enioluwa23) [inline src/chariot/database/Database.py]: Shouldn't this class be called `DatabaseWriter`? Since we are never reading from the database its name should describe the limitations of its responsibility.

File: src/chariot/database/Database.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        Delete the database connection
+        """
+        pass
+
+    @abc.abstractmethod
+    def insertOne(self, dataPoint: dict):
+        """
+        Check for validity of dataPoint, then insert into database
+        """
+        if sorted(dataPoint.keys()) != sorted(['relative_time', 'freeform']):
```
Comment (enioluwa23) [inline src/chariot/database/Database.py]: This seems like an awkward way to check for these. The way this is done, one cannot tell which field was invalid or missing. You can define the `set` of required fields as a class variable and just iterate through the keys to see if
1. Every key in the dict is in the required fields
2. No other key apart from the required ones exist in the dict

Iterating will allow you to know exactly where this dict was flagged as invalid, and you can throw an error accordingly.

File: src/chariot/database/Database.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import abc
+from typing import List
+
+
+class Database(metaclass=abc.ABCMeta):
+    def __init__(self, db_path: str = '../data/iot_database.db'):
```
Comment (enioluwa23) [inline src/chariot/database/Database.py]: Why would we want a default variable here? Users cannot store files directly on the server and we are not providing default databases.

File: src/chariot/database/MongoDatabase.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    def connectToDB(self):
+        # Make the data directory if it does not exist
+        try:
+            os.makedirs(self.db_path)
+        except FileExistsError:
+            pass
+        os.chmod(self.db_path, 0o700)
+
+        self.conn = TinyMongoClient(self.db_path)
```
Comment (enioluwa23) [inline src/chariot/database/MongoDatabase.py]: In general, it seems you are not adhering to our use of type hints. All instance variables should be defined and typed in the constructor. Try running the mypy tool on each of these files and fix the errors that come up.

File: src/chariot/database/MongoDatabase.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    def deleteDB(self):
+        shutil.rmtree(self.db_path)
+
+    def disconnectFromDB(self):
+        self.conn = None
+
+    def insertOne(self, dataPoint: dict):
+        Database.insertOne(self, dataPoint)
+        db_insertion_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
```
Comment (enioluwa23) [inline src/chariot/database/MongoDatabase.py]: Why do we want the date stored in this format? Is there any issue with milliseconds since the epoch?

File: src/chariot/database/SQLiteDatabase.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            'CREATE TABLE IF NOT EXISTS IOTD('
+            'id INTEGER PRIMARY KEY, ' +
+            'db_insertion_time DATETIME DEFAULT CURRENT_TIMESTAMP, ' +
+            'relative_time BIGINT, ' +
+            'freeform TEXT)')
+        self.conn.commit()
+
+    def connectToDB(self):
+        # Make the data directory if it does not exist
+        dirname = os.path.dirname(self.db_path)
```
Comment (enioluwa23) [inline src/chariot/database/SQLiteDatabase.py]: This cannot be the flow. We can't store the user's database locally on our server. The user should provide a remote link to the database through which we connect.

File: src/chariot/database/SQLiteDatabase.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        # Create list of (relative_time, freeform) tuples for every dataPoint
+        insertList = [(dataPoint['relative_time'], dataPoint['freeform'])
+                      for dataPoint in dataPoints]
+        cursor = self.conn.cursor()
+        cursor.executemany(
+            'INSERT INTO IOTD (relative_time, freeform) VALUES (?, ?)',
+            insertList
+        )
+
+    def flush(self):
```
Comment (enioluwa23) [inline src/chariot/database/SQLiteDatabase.py]: Not sure why we need a flush method. These are database writers, they should only write to and not read from or delete from the database, since we do not own it.

File: src/chariot/database/SQLiteDatabase.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        cursor = self.conn.cursor()
+        cursor.execute('DELETE FROM IOTD')
+        self.conn.commit()
+
+    def toCSV(self, outfile='../data/iot_database.csv'):
```
Comment (enioluwa23) [inline src/chariot/database/SQLiteDatabase.py]: This should return a csv string, not write directly to a file.

Comment (enioluwa23) [reply src/chariot/database/SQLiteDatabase.py]: In fact, I think this kind of violates the fact that it is a database writer. In order to convert it to csv it has to read from the database and so I'm not sure why our system would need this method.

File: src/chariot/test/mongo-db.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+from chariot.core.database.MongoDatabase import MongoDatabase
+
+mongo = MongoDatabase()
```
Comment (enioluwa23) [inline src/chariot/test/mongo-db.py]: Please add some structured tests if you want this file in master.

File: src/chariot/database/DatabaseWriter.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import abc
+from typing import List
+
+
+def checkDataPoint(dataPoint: dict):
```
Comment (enioluwa23) [inline src/chariot/database/DatabaseWriter.py]: I see no reason why this is not an instance method of the `DatabaseWriter` class.

**Comment by jh3377** [comment]
The GUI code seems to have been removed again.

File: src/requirements.txt [hunk 12]
```diff
@@ -32,8 +33,8 @@ pycodestyle==2.5.0
 pyflakes==2.1.1
 PyHamcrest==2.0.0
 pylint==2.4.4
+pymongo==3.10.1
 pyparsing==2.4.6
-pytest==5.3.2
```
Comment (enioluwa23) [inline src/requirements.txt]: Why did you remove pytest from the requirements doc?

File: src/chariot/test/mongo-db.py (hunk 10 above)
Comment (enioluwa23) [reply src/chariot/test/mongo-db.py]: This comment still stands. There is no use of pytest here and as such this is not a valid test file.

File: src/chariot/database/MongoDatabase.py (hunk 6 above)
Comment (enioluwa23) [reply src/chariot/database/MongoDatabase.py]: This was not addressed at all.

File: src/chariot/api/network/NetworkManagerAPI.py (hunk 1 above)
Comment (enioluwa23) [reply src/chariot/api/network/NetworkManagerAPI.py]: Why did you delete this file?

File: src/chariot/database/DatabaseWriterFactory.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class DatabaseWriterFactory:
+    def GetDatabaseWriter(DatabaseWriterType: str, connectionString: str) -> DatabaseWriter:
+        if DatabaseWriterType == 'MONGODB':
+            return MongoDatabaseWriter(connectionString)
+        elif DatabaseWriterType == 'MYSQL':
+            return MySQLDatabaseWriter(connectionString)
+        else:
+            return None
```
Comment (enioluwa23) [inline src/chariot/database/DatabaseWriterFactory.py]: This should throw an error, not return `None`.

File: src/chariot/database/MongoDatabaseWriter.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        # Add database insertion time to dataPoint
+        dataPoint['db_insertion_time'] = int(
+            round(time.time() * 1000))  # Millis since epoch
+        self.iot_database.insert_one(dataPoint)
+
+    def insertMany(self, dataPoints: List[dict]):
+        for dataPoint in dataPoints:
+            DatabaseWriter.checkDataPoint(dataPoint)
+            # Add database insertion time to dataPoint
+            dataPoint['db_insertion_time'] = datetime.now().strftime(
```
Comment (ryanhassing) [inline src/chariot/database/MongoDatabaseWriter.py]: Fix this, time should be same form as for insertOne

File: src/chariot/database/MySQLDatabaseWriter.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
+from typing import List, Tuple
+import mysql.connector as connector
+
+from chariot.database.DatabaseWriter import DatabaseWriter
+
+
+class MySQLDatabaseWriter(DatabaseWriter):
+    def __init__(self, connectionString: str):
```
Comment (Juan-Garcia96) [inline src/chariot/database/MySQLDatabaseWriter.py]: make connectionString more configurable, also think about how choosing a table would work in the configuration.

File: src/chariot/database/MySQLDatabaseWriter.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+    def disconnect(self):
+        self.conn.close()
+
+    def initializeTable(self):
+        # Create a new database, to be safe
+        self.cursor.execute("CREATE DATABASE IF NOT EXISTS iot_data;")
+        self.cursor.execute("USE iot_data;")
+        self.cursor.execute(
+            "CREATE TABLE IF NOT EXISTS data(id INTEGER PRIMARY KEY AUTO_INCREMENT, db_insertion_time TIMESTAMP DEFAULT CURRENT_TIMESTAMP, relative_time BIGINT, freeform VARBINARY(64535))"
```
Comment (ryanhassing) [inline src/chariot/database/MySQLDatabaseWriter.py]: Should connect to existing database/table if specified by user

File: src/chariot/database/MySQLDatabaseWriter.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        self.conn: connector.connection.MySQLConnection = connector.connect(
+            user=_user, password=_password, host=_host, port=_port)
+        self.cursor: connector.cursor.MySQLCursor = self.conn.cursor()
+
+    def disconnect(self):
+        self.conn.close()
+
+    def initializeTable(self):
+        # Create a new database, to be safe
+        self.cursor.execute("CREATE DATABASE IF NOT EXISTS iot_data;")
```
Comment (Juan-Garcia96) [inline src/chariot/database/MySQLDatabaseWriter.py]: I see the method declared, but is it being used?

File: src/chariot/database/MySQLDatabaseWriter.py (hunk 15 above)
Comment (enioluwa23) [reply src/chariot/database/MySQLDatabaseWriter.py]: Turn to the way it's done for devices on how to properly implement configurations.

File: src/chariot/utility/AbstractFactory.py [hunk 18]
```diff
@@ -1,16 +1,18 @@
...
         # override this for more descriptive error messages
         self.instanceName: str = 'instance'
 
     def getInstance(self, config: Type[Configuration]) -> Any:
-        instanceType: str = config['type']
+        instanceType: str = config.type
         if instanceType not in self.instanceMap:
-            raise AssertionError(f'Unsupported {self.instanceName} type "{instanceType}"')
+            raise AssertionError(
+                'Unsupported {self.instanceName} type "{instanceType}"')
```
Comment (ryanhassing) [inline src/chariot/utility/AbstractFactory.py]: Change back.
